Extension/src/api.ts

```diff
@@ -0,0 +1,25 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+
+import * as vscode from 'vscode';
+
+export interface CppToolsApi {
+    registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void;
+}
+
+export interface SourceFileConfiguration {
+    includePaths: string[];   // This must also include the system include path (compiler defaults)
+    defines: string[];        // This must also include the compiler default defines (__cplusplus, etc)
+    intelliSenseMode: string; // "msvc-x64" or "clang-x64"
+    forcedInclude?: string[]; // Any files that need to be included before the source file is parsed
+    standard: string;         // The C or C++ standard
+}
+ 
+export interface CustomConfigurationProvider {
+    name: string;
+    canProvideConfiguration(uri: vscode.Uri): boolean;
+    provideConfiguration(uri: vscode.Uri): Thenable<SourceFileConfiguration>;
+}
```

Comment:
Add method for cpptools to request updated custom configs for tracked files.  e.g. `provideConfigurations(uris: vscode.Uri[]): Thenable<SourceFileConfiguration[]>`

Consider removing `provideConfiguration` in favor of this newer general purpose API.  This would mean you need to update client.ts to send the batch instead of a single config at a time.

Reply:
Also add `vscode.Event` property that cpptools can subscribe to and listen for configuration change events. e.g. `didChangeCustomConfiguration: vscode.EventEmitter<T>` (can T = void?)

If this doesn't work, the alternative is to add `didChangeCustomConfigurations(provider: CustomConfigurationProvider)` to the CppToolsApi interface.  Either way is ok.